### PR TITLE
Added support for IQueryExpressionInterceptor, IMemberTranslator via DataOptions.

### PR DESCRIPTION
**Pull request by sdanyliv**
Replacement for https://github.com/linq2db/linq2db/pull/3510

### CODE REVIEW

File: Source/LinqToDB/Data/CommandInfo.cs [hunk 1]
```diff
@@ -330,10 +331,10 @@ IEnumerable<T> ReadEnumerator<T>(DataReaderWrapper rd, IExecutionScope? scope, b
 
 						DbDataReader reader;
 
-						if (((IDataContext)DataConnection).UnwrapDataObjectInterceptor is {} interceptor)
+						if (DataConnection is IInterceptable<IUnwrapDataObjectInterceptor> { Interceptor: not null } interceptable)
```
Comment (viceroypenguin) [inline Source/LinqToDB/Data/CommandInfo.cs:334]: same for others
```suggestion
if (DataConnection is IInterceptable<IUnwrapDataObjectInterceptor> { Interceptor: { } interceptor })
{
  using (...)
    reader = interceptor...
}
```

File: Source/LinqToDB/Remote/RemoteDataContextBase.Interceptors.cs [hunk 2]
```diff
@@ -10,7 +10,8 @@ public abstract partial class RemoteDataContextBase :
 		IInterceptable<IDataContextInterceptor>,
 		IInterceptable<IEntityServiceInterceptor>,
 		IInterceptable<IUnwrapDataObjectInterceptor>,
-		IInterceptable<IEntityBindingInterceptor>
+		IInterceptable<IEntityBindingInterceptor>,
+		IInterceptable<IQueryExpressionInterceptor>
```
Comment (viceroypenguin) [inline Source/LinqToDB/Remote/RemoteDataContextBase.Interceptors.cs:14]:
```suggestion
		IInterceptable<IQueryExpressionInterceptor>,
```

File: Source/LinqToDB/Data/CommandInfo.cs (hunk 1 above)
Comment (sdanyliv) [reply Source/LinqToDB/Data/CommandInfo.cs:334]: Good point will fix that.

**Review by viceroypenguin — APPROVED** [review]

File: Source/LinqToDB/Data/DataConnection.Async.cs [hunk 3]
```diff
@@ -138,18 +139,18 @@ public async Task EnsureConnectionAsync(CancellationToken cancellationToken = de
 
 				if (_connection.State == ConnectionState.Closed)
 				{
-					if (_connectionInterceptor != null)
+					if (((IInterceptable<IConnectionInterceptor>)this).Interceptor != null)
```
Comment (viceroypenguin) [inline Source/LinqToDB/Data/DataConnection.Async.cs:142]: could you do:
```cs
if (this is IInterceptable<IConnectionInterceptor> { Interceptor: { } interceptor })
```
like the other pattern matching ones?

File: Source/LinqToDB/Common/OptionsContainer.cs [hunk 4]
```diff
@@ -116,13 +116,5 @@ public virtual TSet Get<TSet>()
 
 			return set;
 		}
-
-		protected void Apply<TA>(TA obj)
```
Comment (igor-tkachev) [inline Source/LinqToDB/Common/OptionsContainer.cs:120]: What's wrong with this code?

**Review by igor-tkachev — CHANGES_REQUESTED** [review]
Is this PR about Interceptors or Options? I see a lot of questionable Options refactoring. Can you split it, so we will be able to discuss it separately?

File: Source/LinqToDB/Common/OptionsContainer.cs (hunk 4 above)
Comment (sdanyliv) [reply Source/LinqToDB/Common/OptionsContainer.cs:120]: Tight coupling with DataConnection, DataContext, RemoteDataContext. Options should know nothing about these classes. Also current code has strange unpredictable side effects. 

Removing all of this strange stuff made code readable, simpler, predictable and performant.

Comment (igor-tkachev) [reply Source/LinqToDB/Common/OptionsContainer.cs:120]: Then split this PR, so it will be easier to revert it later, when I will need to extend it.

Comment (sdanyliv) [reply Source/LinqToDB/Common/OptionsContainer.cs]: Reverted this change, will create new PR for that after merging with `version_6` branch.

**Review by MaceWindu — APPROVED** [review]

**Pull request merged**